Replace Fabric Crashlytics with Sentry

Gudahtt opened the pull request:
* Remove Fabric and Crashlytics
* Add Sentry reporting
  Errors are reported using `captureException`, and non-errors are added as breadcrumbs. This means the non-error events won't be directly sent to Sentry, but they will be included as context for any errors that do occur.

  The sentry configuration is mainly in the `sentry.debug.properties` file. The only config outside of that file is the DSN used for reporting errors, which is hard-coded in the `setupSentry.js` module.

  There are three separate sets of config: default, debug, and release. The default config is missing the auth token, so it can't be used to publish source maps. However, builds that use the default config will still correctly report errors to the `test-metamask-mobile` Sentry project.

  The debug config is identical to the default config except that it includes the auth token. The debug config is for publishing any non-production builds (e.g. testing, pre-releases).

  The release configuration is to be used for production builds only. This is the only config that sends errors to the `metemask-mobile` Sentry project.

  The debug and release config files have not been added to the repo. They are automatically instantiated from the example files if the `MM_SENTRY_AUTH_TOKEN` environment variable is set. Setting this environment variable in CI should allow publishing from CI.

Closes #1321

Gudahtt committed:
Remove Fabric and Crashlytics

The Logger now logs to `console.log` exclusively.

Gudahtt committed:
Add Sentry reporting

Errors are reported using `captureException`, and non-errors are added
as breadcrumbs. This means the non-error events won't be directly sent
to Sentry, but they will be included as context for any errors that do
occur.

The sentry configuration is mainly in the `sentry.debug.properties`
file. The only config outside of that file is the DSN used for
reporting errors, which is hard-coded in the `setupSentry.js` module.

There are three separate sets of config: default, debug, and release.
The default config is missing the auth token, so it can't be used to
publish source maps. However, builds that use the default config will
still correctly report errors to the `test-metamask-mobile` Sentry
project.

The debug config is identical to the default config except that it
includes the auth token. The debug config is for publishing any
non-production builds (e.g. testing, pre-releases).

The release configuration is to be used for production builds only.
This is the only config that sends errors to the `metemask-mobile`
Sentry project.

The debug and release config files have not been added to the repo.
They are automatically instantiated from the example files if the
`MM_SENTRY_AUTH_TOKEN` environment variable is set. Setting this
environment variable in CI should allow publishing from CI.

Closes #1321

Gudahtt commented on android/sentry.properties:
```diff
@@ -0,0 +1,5 @@
+defaults.url=https://sentry.io/
+defaults.org=metamask
+defaults.project=metamask-mobile
+auth.token=f3a55b2e27454dd4aa1628a7ee7e71ae7e70ca0089e24ffeab8c5d6ae6f47369
```
This should be replaced with a new key that is added via environment file or variable.

Ideally we can leave a public key for a "test" account maybe, as we do for the extension? :man_shrugging:

Gudahtt commented on app/util/setupSentry.js:
```diff
@@ -0,0 +1,20 @@
+
+const METAMASK_ENVIRONMENT = process.env['METAMASK_ENVIRONMENT']; // eslint-disable-line dot-notation
+// const SENTRY_DSN = 'https://ae39e4b08d464bba9fbf121c85ccfca0@sentry.io/2299799'
+const TEST_SENTRY_DSN = 'https://1cb0be9b81554de381b65ffa80874924@sentry.io/2453959';
```
Whoops, this was from my own testing with a throwaway account - I'll remove this.

Gudahtt commented on app/util/setupSentry.js:
```diff
@@ -0,0 +1,20 @@
+import { init } from '@sentry/react-native';
+import { Dedupe, ExtraErrorData } from '@sentry/integrations';
+
+const METAMASK_ENVIRONMENT = process.env['METAMASK_ENVIRONMENT']; // eslint-disable-line dot-notation
```
This environment isn't set anywhere yet

rickycodes commented on android/app/src/main/java/io/metamask/MainApplication.java:
```diff
@@ -50,10 +50,10 @@ public boolean getUseDeveloperSupport() {
 		protected List<ReactPackage> getPackages() {
 		return Arrays.<ReactPackage>asList(
 				new MainReactPackage(),
+            new RNSentryPackage(),
```
nit: seems tabbing is off here?

Gudahtt replied:
Thanks for the reminder - I had been meaning to fix this. Some combination of the automatic linker and my brand-new environment has left a lot of the formatting in a bad state.

Gudahtt commented on app/util/Logger.js:
```diff
@@ -18,14 +17,15 @@ export default class Logger {
 		} else if (metricsOptIn === 'agreed') {
-			Fabric.Crashlytics.log(JSON.stringify(args));
+			addBreadcrumb({
+				message: JSON.stringify(args)
```
There is a lot more useful info we could add here to the breadcrumb, but this is a good start at least. I might improve this later by changing the arguments to allow logging structured data, rather than an arbitrary array

Gudahtt commented:
To test this locally, I added the line `throw new Error('TEST')` to the constructor of `app/components/Views/Root/index.js`. It should then blow up immediately when the app opens.

You can test the breadcrumbs by adding a `Logger.log` call as well, right above the error.

danjm commented on sentry.properties:
```diff
@@ -0,0 +1,5 @@
+defaults.url=https://sentry.io/
+defaults.org=metamask
+defaults.project=test-metamask-mobile
+auth.token=
```
How does the auth token work? When is it needed?

Gudahtt replied:
It is used to publish sourcemaps, which happens as part of the release build

Gudahtt replied:
A similar command is in place for the extension (`sentry:publish`) which also needs a token; any submission of release artifacts requires this token.

Unfortunately with React Native, this command is integrated with the normal build process; I'm not sure how to pull it into a separate step. The modifications to the build process were done automatically when linking `@sentry/react-native`, and I wasn't eager to customize them. At least it's automatically skipped in dev mode though, so it shouldn't affect dev environments

danjm commented on app/util/Logger.js:
```diff
@@ -18,14 +17,15 @@ export default class Logger {
 		if (__DEV__) {
 			args.unshift('[MetaMask DEBUG]:');
 			console.log.apply(null, args); // eslint-disable-line no-console
 		} else if (metricsOptIn === 'agreed') {
```
Does mobile metrics opt-in include a note about error reporting?

Gudahtt replied:
Not sure. I was curious about this as well; on the extension we don't ask for permission to collect error information at all.

danjm commented:
I've completed code review.

- sentry api usage seems correct, and replacement of fabric calls seem correct
- javascript code looks sound
- code related to fabric and crashlytics has been deleted
- build.sh looks fine to my not so bash trained eyes

Gudahtt committed:
Use absolute path for Sentry file

The SENTRY_PROPERTIES environment variable is used at different
directory levels in the Android and iOS builds respectively; it's
used one level down with iOS, but two levels down for Android. This
means the properties path used is incorrect on iOS at the moment.

Instead an absolute path is now used, to prevent any such errors in the
future.

The error message for the missing auth token has been improved as well.

Gudahtt committed:
Improve handling of missing auth token

The properties file is now checked to ensure the auth token is present,
so the build can fail early rather than partway through.

The logic for handling the auth token has been moved to a separate
function as well, so it can be shared between the release and
prerelease builds.

Gudahtt committed:
Force setting METAMASK_ENVIRONMENT explicitly for release builds

The Sentry environment is set to `local` by default (e.g. for dev
builds), but the build script now requires it to be explicitly set for
release builds. This is to ensure it isn't missed in the production
builds, which are done manually.

Gudahtt committed:
Ensure only Errors are thrown/rejected

Whenever using the `throw` keyword or calling `reject` in a Promise,
the value passed should be an Error. Various tools expect thrown
"errors" to be errors; this is the convention. Using Error types is
also more useful for debugging, as they have stack traces.

Gudahtt committed:
Update `Logger.error` function signature

`Logger.error` now expects an actual error object as the first
argument. This allows us to submit proper Errors with stack traces to
Sentry. The second parameter is an optional set of extra data, which
gets attached to the Error and is viewable on Sentry in the "Additional
Data" section of the error report.

Unfortunately any messages added to contextualize errors had to be
submitted as extra data instead of as a wrapper Error. Wrapper error
types provide for a richer debugging experience in general, because
you get additional context for the error plus an additional stack trace
to help trace the flow of the error. Sentry even has an Integration
meant to facilitate this pattern, the `LinkedErrors` integration.
But unfortunately that integration doesn't seem to work with React
Native, so that option was off the table.

Gudahtt committed:
Update `@sentry` packages to latest

The most recent Sentry packages have better support for capturing
native crashes on Android. I had held off on this update at first
beceause I thought it required React Native >= 0.60, but apparently
not. It still works on 0.59, just not as well (no sourcemaps)

Gudahtt committed:
Fix bug with build script

It was expecting an alphabetical key instead of an alphanumeric one.

estebanmino reviewed (approved):
working on IOS! 🥇

Gudahtt committed:
Merge branch 'develop' into sentry
